app/vmselect/prometheus/prometheus.go

```diff
@@ -613,8 +617,12 @@ func LabelsHandler(qt *querytracer.Tracer, startTime time.Time, at *auth.Token,
 	if err != nil {
 		return err
 	}
+	minLimit := *maxUniqueTimeseries
+	if limit > 0 && limit < *maxUniqueTimeseries {
+		minLimit = limit
+	}
 	denyPartialResponse := searchutils.GetDenyPartialResponse(r)
-	sq := storage.NewSearchQuery(at.AccountID, at.ProjectID, cp.start, cp.end, cp.filterss, *maxUniqueTimeseries)
+	sq := storage.NewSearchQuery(at.AccountID, at.ProjectID, cp.start, cp.end, cp.filterss, minLimit)
```

Comment:
This limit may work incorrectly due to the same logic explained for `/api/v1/label/labelName/values` handler.